File: internal/identifiers/identifiers.go
// Package identifiers 读取系统 user/group/cgroup 列表，对齐原 include.uc 的 get_users/get_groups/get_cgroups。
package identifiers

import (
	"bufio"
	"os"
	"path/filepath"
	"strings"
)

type Identifiers struct {
	Users      []string `json:"users"`
	Groups     []string `json:"groups"`
	Cgroups    []string `json:"cgroups"`
	OSType     string   `json:"os_type"` // "openwrt" 或 "linux"
}

// Get 收集系统标识符。
func Get() *Identifiers {
	osType := detectOSType()
	return &Identifiers{
		Users:   getUsers(),
		Groups:  getGroups(),
		Cgroups: getCgroups(),
		OSType:  osType,
	}
}

// detectOSType 检测系统类型：OpenWrt 或普通 Linux。
// 判定依据：/etc/openwrt_release 存在即为 OpenWrt。
func detectOSType() string {
	if _, err := os.Stat("/etc/openwrt_release"); err == nil {
		return "openwrt"
	}
	return "linux"
}

func getUsers() []string {
	f, err := os.Open("/etc/passwd")
	if err != nil {
		return nil
	}
	defer f.Close()
	var out []string
	sc := bufio.NewScanner(f)
	for sc.Scan() {
		line := sc.Text()
		if i := strings.IndexByte(line, ':'); i > 0 {
			out = append(out, line[:i])
		}
	}
	return out
}

func getGroups() []string {
	f, err := os.Open("/etc/group")
	if err != nil {
		return nil
	}
	defer f.Close()
	var out []string
	sc := bufio.NewScanner(f)
	for sc.Scan() {
		line := sc.Text()
		if i := strings.IndexByte(line, ':'); i > 0 {
			out = append(out, line[:i])
		}
	}
	return out
}

// getCgroups 对齐 include.uc：cgroup v2 时遍历 /sys/fs/cgroup 子目录，排除 services/proxy。
func getCgroups() []string {
	if !isCgroupV2() {
		return nil
	}
	const root = "/sys/fs/cgroup/"
	var out []string
	_ = filepath.Walk(root, func(path string, info os.FileInfo, err error) error {
		if err != nil || !info.IsDir() {
			return nil
		}
		rel := strings.TrimPrefix(path, root)
		if rel == "" {
			return nil
		}
		if rel == "services/proxy" {
			return nil
		}
		out = append(out, rel)
		return nil
	})
	return out
}

// isCgroupV2 对齐 include.uc get_cgroups_version()：mount 输出含 "^cgroup" 视为 v1，否则 v2。
func isCgroupV2() bool {
	f, err := os.Open("/proc/mounts")
	if err != nil {
		return true
	}
	defer f.Close()
	sc := bufio.NewScanner(f)
	for sc.Scan() {
		fields := strings.Fields(sc.Text())
		if len(fields) >= 3 && fields[2] == "cgroup" {
			return false
		}
	}
	return true
}

File: src/identifiers.rs
//! 读取系统 user/group/cgroup 列表，对齐原 include.uc 的 get_users/get_groups/get_cgroups。

use std::fs;
use std::path::Path;

use serde::Serialize;

#[derive(Debug, Serialize)]
pub struct Identifiers {
    #[serde(default)]
    pub users: Vec<String>,
    #[serde(default)]
    pub groups: Vec<String>,
    #[serde(default)]
    pub cgroups: Vec<String>,
    pub os_type: String, // "openwrt" 或 "linux"
}

/// Get 收集系统标识符。
pub fn get() -> Identifiers {
    Identifiers {
        users: get_users(),
        groups: get_groups(),
        cgroups: get_cgroups(),
        os_type: detect_os_type(),
    }
}

fn detect_os_type() -> String {
    if Path::new("/etc/openwrt_release").exists() {
        "openwrt".to_string()
    } else {
        "linux".to_string()
    }
}

fn get_users() -> Vec<String> {
    let mut out = vec![];
    if let Ok(data) = fs::read_to_string("/etc/passwd") {
        for line in data.lines() {
            if let Some(i) = line.find(':') {
                if i > 0 {
                    out.push(line[..i].to_string());
                }
            }
        }
    }
    out
}

fn get_groups() -> Vec<String> {
    let mut out = vec![];
    if let Ok(data) = fs::read_to_string("/etc/group") {
        for line in data.lines() {
            if let Some(i) = line.find(':') {
                if i > 0 {
                    out.push(line[..i].to_string());
                }
            }
        }
    }
    out
}

/// getCgroups 对齐 include.uc：cgroup v2 时遍历 /sys/fs/cgroup 子目录，排除 services/proxy。
fn get_cgroups() -> Vec<String> {
    if !is_cgroup_v2() {
        return vec![];
    }
    let root = "/sys/fs/cgroup/";
    let mut out = vec![];
    walk_dir(root, root, &mut out);
    out
}

fn walk_dir(root: &str, path: &str, out: &mut Vec<String>) {
    let entries = match fs::read_dir(path) {
        Ok(e) => e,
        Err(_) => return,
    };
    for e in entries.flatten() {
        let p = e.path();
        if !p.is_dir() {
            continue;
        }
        let rel = path.strip_prefix(root).unwrap_or("");
        let name = e.file_name().to_string_lossy().to_string();
        let full = if rel.is_empty() {
            name.clone()
        } else {
            format!("{}/{}", rel, name)
        };
        if full == "services/proxy" {
            // 仍递归子目录，但跳过该条目本身
        } else {
            out.push(full.clone());
        }
        let child = format!("{}/{}", path, name);
        walk_dir(root, &child, out);
    }
}

fn is_cgroup_v2() -> bool {
    let f = match fs::read_to_string("/proc/mounts") {
        Ok(s) => s,
        Err(_) => return true,
    };
    for line in f.lines() {
        let fields: Vec<&str> = line.split_whitespace().collect();
        if fields.len() >= 3 && fields[2] == "cgroup" {
            return false;
        }
    }
    true
}
